### PR TITLE
Bump up nilearn required dependencies versions

### PR DESCRIPTION
**Pull request by lesteve**
Supersedes #397.

The main rationale was for the versions to be roughly 2 years old and
to support neurodebian on Ubuntu 12.04.

As part of this change, implemented a minimum required version check for
matplotlib in `nilearn/plotting/__init__.py`.

The ubuntu and ubuntu-no-matplotlib travis distribs have been replaced by conda based ones using the same package versions available as in Ubuntu 14.04.

### CODE REVIEW

File: nilearn/plotting/__init__.py [hunk 1]
```diff
@@ -3,26 +3,31 @@
...
+            matplotlib.use('Agg')
     except ImportError:
-        pass
         # No need to fail here, eg during tests
+        pass
+    else:
+        # When matplotlib was successfully imported we need to check
+        # that the version is greater that the minimum required one
+        _import_module_with_version_check('matplotlib',
+                                          OPTIONAL_MATPLOTLIB_MIN_VERSION)
```
Comment (lesteve) [inline nilearn/plotting/__init__.py]: I had to shuffle things around a little bit here. Basically I both wanted
* the matplotlib minimum version should be always checked 
* Agg backend should only be used for posix platforms when the DISPLAY environment variable is not defined (e.g. on Travis)

File: README.rst [hunk 2]
```diff
@@ -36,16 +36,14 @@ The required dependencies to use the software are:
 
 * Python >= 2.6,
 * setuptools
-* Numpy >= 1.6
+* Numpy >= 1.6.1
 * SciPy >= 0.9
 * Scikit-learn >= 0.12.1
-* Nibabel >= 1.1.0.
+* Nibabel >= 1.3.0
```
Comment (GaelVaroquaux) [inline README.rst]: Why?

File: nilearn/plotting/__init__.py (hunk 1 above)
Comment (GaelVaroquaux) [inline nilearn/plotting/__init__.py]: Did you check that the version on Ubuntu 12.04 passes the test? It seems to be an 1.1.1rc1 from what I can see here:
http://packages.ubuntu.com/search?suite=precise&searchon=names&keywords=matplotlib

File: README.rst (hunk 2 above)
Comment (lesteve) [inline README.rst]: 1.3.0 is quite old according to piprot:

```
~/dev/nilearn $ piprot <(echo nibabel==1.3.0)
nibabel (1.3.0) is 819 days out of date. Latest is 2.0.0
Your requirements are 819 days out of date
```

Also this is the version that neurodebian on Ubuntu 12.04 comes with.

Comment (AlexandreAbraham) [inline README.rst]: Note: we don't support 12.04 anymore (because of numpy / scipy). We only support 12.04 neurodebian. This way, we can upgrade most of packages.
From 1.1.0 to 1.3.0, a lot of significant bugs were fixed. Also, some code to handle affine transformations was added. We can rely on it and thus simplify nilearn's code.

File: nilearn/plotting/__init__.py (hunk 1 above)
Comment (lesteve) [inline nilearn/plotting/__init__.py]: Yeah LooseVersion doesn't seem to be very knowledgeable about rc, alpha, et al. versions ...

```
In [1]: from distutils.version import LooseVersion

In [2]: LooseVersion('1.1.1rc') > LooseVersion('1.1.1')
Out[2]: True

In [3]: LooseVersion('1.1.1.a1') > LooseVersion('1.1.1')
Out[3]: True
```

Also we decided to drop support for bog-standard Ubuntu 12.04 versions long ago in #397 by requiring sklearn >= 0.12.1.

File: README.rst (hunk 2 above)
Comment (GaelVaroquaux) [inline README.rst]: > From 1.1.0 to 1.3.0, a lot of significant bugs were fixed. Also, some code to
> handle affine transformations was added. We can rely on it and thus simplify
> nilearn's code.

OK, these are good arguments.

Comment (AlexandreAbraham) [inline README.rst]: However, I think we should do that in a separate PR where we actually use these new features. In the meantime, if the code works with 1.1.0, I see no (good enough) reason to ask for a more recnt one.

Comment (lesteve) [inline README.rst]: > In the meantime, if the code works with 1.1.0, I see no (good enough) reason to ask for a more recnt one.

What about finding some middle ground and going for 1.2.2, which is the version the version of [python-nibabel](http://packages.ubuntu.com/trusty/python-nibabel) in Ubuntu 14.04?

Comment (GaelVaroquaux) [inline README.rst]: > What about finding some middle ground and going for 1.2.2, which is the
> version the version of python-nibabel in Ubuntu 14.04?

We certainly should not go above that.

But what is the argument in favor of bumping the version? What is the
logic? What does it gain us and the user?

Comment (lesteve) [inline README.rst]: Citing from the original PR #397.

>> In particular does scikit-learn 0.12.1 have a special meaning?
> No. It's a configuration that's 2 years old, and we felt that 2 years old was a good time.

I guess I assumed that updating all dependencies in a reasonably uniform manner was the sensible thing to do.

Side comment, it seems that with our installation instructions (pip install nilearn -U --user) nibabel will always been updated to the latest version anyway (I tried both with nibabel is install via apt-get and in a conda environment).

Comment (GaelVaroquaux) [inline README.rst]: > Citing from the original PR #397.

>         In particular does scikit-learn 0.12.1 have a special meaning?
>         No. It's a configuration that's 2 years old, and we felt that 2 years
>         old was a good time.

> I guess I assumed that updating all dependencies in a reasonably
> uniform manner was the sensible thing to do.

That means that you are forcing all your users to upgrade their system
every two year, when you decide that it's time (like now). These people
want to use nilearn to solve their problem as fast as possible, and
probably don't fancy having an update to do. We need to bump dependencies
only if there is an argument for it. The old scikit-learn was a big
problem, with accumulating backports. I don't see a strong argument for
the nilearn dependency bump, but maybe there is one.

> Side comment, it seems that with our installation instructions (pip install
> nilearn -U --user) nibabel will always been updated to the latest version
> anyway (I tried both with nibabel is install via apt-get and in a conda
> environment).

Don't get me started on how the semantics of pip are broken. It's the
only thing that we have unfortunately.

Comment (lesteve) [inline README.rst]: OK I reverted nibabel required version to 1.1.0.
